src/sequence.jl

```diff
@@ -1,445 +1,12 @@
-    loglik = 0.0  # remains 0 if no observables
-    for t = 2:T+1
-        u_f[t] = A_1 * u_f[t-1] .+ B * noise[t-1]
-        u[t] = A_0 + A_1 * u[t-1] + quad(A_2, u_f[t-1]) .+ B * noise[t-1]
-        z[t] = C_0 + C_1 * u[t] + quad(C_2, u_f[t])
-        err = observables[t-1] - z[t]
-        loglik += logpdf(D, err)  # z_0 doesn't enter likelihood        
-    end
-    return StateSpaceSolution(copy(z), copy(u), noise, nothing, loglik)
-end
-
-
-# QTI with noise and observables is a joint likelihood
-function _solve(
-    alg::QTILikelihood,
-    noise,
-    observables,
-    A_0,
-    A_1,
-    A_2,
-    B,
-    C_0,
-    C_1,
-    C_2,
-    D,
-    u0,
-    tspan,
-)
-    T = tspan[2]
-    @assert tspan[1] == 0
-    @assert length(noise) == T
-
-    z0 = C_0 + C_1 * u0 + quad(C_2, u0)  # the same u0 used for initial observation
-    u_f = Zygote.Buffer(Vector{typeof(u0)}(undef, T + 1))
-    u = Zygote.Buffer(Vector{typeof(u0)}(undef, T + 1))
-    z = Zygote.Buffer(Vector{typeof(z0)}(undef, T + 1))
-    u[1] = u0
-    u_f[1] = u0
-    z[1] = z0
-    loglik = 0.0  # remains 0 if no observables
-    for t = 2:T+1
-        u_f[t] = A_1 * u_f[t-1] .+ B * noise[t-1]
-        u[t] = A_0 + A_1 * u[t-1] + quad(A_2, u_f[t-1]) .+ B * noise[t-1]
-        z[t] = C_0 + C_1 * u[t] + quad(C_2, u_f[t])
-        err = observables[t-1] - z[t]
-        loglik += logpdf(D, err)  # z_0 doesn't enter likelihood        
-    end
-    return StateSpaceSolution(copy(z), copy(u), noise, nothing, loglik)
-end
-
-
-# QTI with noise and observables is a joint likelihood
-function _solve(
-    alg::QTI,
-    noise,
-    observables::Nothing,
-    A_0,
-    A_1,
-    A_2,
-    B,
-    C_0,
-    C_1,
-    C_2,
-    D,
-    u0,
-    tspan,
-)
-    T = tspan[2]
-    @assert tspan[1] == 0
-    @assert length(noise) == T
-
-    z0 = C_0 + C_1 * u0 + quad(C_2, u0)  # the same u0 used for initial observation
-    u_f = Zygote.Buffer(Vector{typeof(u0)}(undef, T + 1))
-    u = Zygote.Buffer(Vector{typeof(u0)}(undef, T + 1))
-    z = Zygote.Buffer(Vector{typeof(z0)}(undef, T + 1))
-    u[1] = u0
-    u_f[1] = u0
-    z[1] = z0
-    loglik = 0.0  # remains 0 if no observables
-    for t = 2:T+1
-        u_f[t] = A_1 * u_f[t-1] .+ B * noise[t-1]
-        u[t] = A_0 + A_1 * u[t-1] + quad(A_2, u_f[t-1]) .+ B * noise[t-1]
-        z[t] = C_0 + C_1 * u[t] + quad(C_2, u_f[t])
-    end
-    return StateSpaceSolution(copy(z), copy(u), noise, nothing, loglik)
-end
-
-
 ## General function mapping for the Perturbation Soultions
 # ideally, we wouldn't need the specializations for the QTI/etc if these are just as efficient
 
 # First order
-dssm_evolution(x, sol::FirstOrderPerturbationSolution, t) = sol.A * x
-dssm_volatility(x, sol::FirstOrderPerturbationSolution, t) = sol.B
-dssm_observation(x, sol::FirstOrderPerturbationSolution, t) = sol.C * x
+dssm_evolution(u, sol::FirstOrderPerturbationSolution, t) = sol.A * u # f
+dssm_volatility(u, sol::FirstOrderPerturbationSolution, t) = sol.B'    # g
```

Comment:
Why do we have to take a transpose of this? If `B` is `n_u * n_eps` and the noises stacks as a vector of `n_eps` elements, then it should naturally work?

Reply:
Mostly a dimensionality issue to make sure that we handle vector/matrix volatility conformability, but let me confirm that this is still correct.

Reply:
I switched this to be `diagm(vec(sol.B))`, does a diagonal matrix here seem consistent with people's intuition?

Reply:
No, a diagonal matrix here is insufficient. This is basically the loadings of the shocks to the states, and it can be multiple shocks hitting the same equation.

Reply:
So it should be diagonal only if `sol.B` is `n x 1`, otherwise it should be a fully specified matrix?

Reply:
I my question here is about moving from vector loadings (the current way) to matrix loadings (my proposed way).

Reply:
Yes B should be a matrix, as it was before.

Reply:
I mean a multicolumn matrix. Currently it is a single-column matrix which is not conformable in the canonical state space representation.

Reply:
...

Yes, `B` should be a multicolumn matrix. It's not a vector, while it could be a vector if we only have one shock.